### PR TITLE
Lints for allowable key usages as per RFC 8813 Section 3 and RFC 3279 Section 2.3.1

### PR DESCRIPTION
**Pull request by mtgag**
This pull request addresses issues regarding the proper key usage values for specific public key types. This work has been made in close cooperation with the D-Trust CA who also authored https://github.com/zmap/zlint/pull/599.

It has four lints. One is for EC keys that is based on RFC 8813 and the pull request https://github.com/zmap/zlint/pull/599. Incorporating 599 to this PR has been coordinated with the authors to adress the key usages issues as a whole.

Three lints regard RSA keys. One lint is for subscriber certificates, one for disallowed key usage values in CA certificates and the third for disallowed key usage values in CA certificates when certain values are present. These lints are based on RFC 3279.

These lints are motivated by a few incidents in the past regarding proper values of key usage. We would be grateful if you could incorporate this PR in the main project. We believe that this would be a benefit for the project and the CAs that use zlint to help them avoid issues in the future.

### CODE REVIEW

**Review by christopher-henderson — CHANGES_REQUESTED** [review]
Thank you for the PR!

In addition to the questions I had regarding whether these should actually be warnings, would you mind also adding the following functions to [util/ku.go](https://github.com/zmap/zlint/blob/master/v3/util/ku.go) and using them?

```go
// HasKeyUsageOID returns whether-or-not the OID 2.5.29.15 is present in the given certificate's extensions.
func HasKeyUsageOID(c *x509.Certificate) bool {
	return IsExtInCert(c, KeyUsageOID)
}

// HasKeyUsage returns whether-or-not the given x509.KeyUsage is present within the
// given certificate's KeyUsage bitmap. The certificate, however, is NOT checked for
// whether-or-not it actually has a key usage OID. If you wish to check for the presence
// of the key usage OID, please use HasKeyUsageOID.
func HasKeyUsage(c *x509.Certificate, usage x509.KeyUsage) bool {
	return KeyUsageIsPresent(c.KeyUsage, usage)
}

// KeyUsageIsPresent checks the provided bitmap (keyUsages) for presence of the provided x509.KeyUsage.
func KeyUsageIsPresent(keyUsages x509.KeyUsage, usage x509.KeyUsage) bool {
	return keyUsages&usage != 0
}
```

It's just a bit of hygiene to help reduce the number of free-floating bitmasks (which can obfuscate what is actually going on if one is not careful).

While in that file, you will also find [KeyUsageToString](https://github.com/zmap/zlint/blob/74f454196357f798ca087df6d43e80d9a7a4debd/v3/util/ku.go#L7) which may be helpful for constructing our `Details` strings.

File: v3/lints/rfc/lint_rsa_allowed_ku_ca.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+   following values MAY be present:
+
+      digitalSignature;
+      nonRepudiation;
+      keyEncipherment;
+      dataEncipherment;
+      keyCertSign; and
+      cRLSign.
+
+   However, this specification RECOMMENDS that if keyCertSign or cRLSign
```
[the remark this answers is not in the document]
Comment (mtgag) [reply v3/lints/rfc/lint_rsa_allowed_ku_ca.go]: done.

**Comment by mtgag** [comment]
Just made a commit that adresses the code enhancements issues. Also added the missing lint for KU for ECC keys. This is also related to lint_ecdsa_ee_invalid_ku.go. We may want to keep one lint.

The Error/Warn part is going to be addressed in a separate commit.

File: v3/lints/rfc/lint_rsa_allowed_ku_ca.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	if c.KeyUsage&x509.KeyUsageDecipherOnly != 0 {
+		invalidKUs = append(invalidKUs, "decipherOnly")
+	}
+
+	if len(invalidKUs) > 0 {
+		// Sort the invalid KUs to allow consistent ordering of Details messages for unit testing
+		sort.Strings(invalidKUs)
+		return &lint.LintResult{
+			Status:  lint.Warn,
```
[the remark this answers is not in the document]
Comment (mtgag) [reply v3/lints/rfc/lint_rsa_allowed_ku_ca.go:88]: In the context of CA compliance with certain policies I believe this is answered as: other key usages MUST NOT be present (error).  A full-text case sensitive search for the word MAY in zlint gives some hints about how this is handled. In lint_ecdsa_ee_invalid_ku.go this is as answered as: other key usages SHOULD NOT be present (notice), (newer specifications (RFC 8813) are more explicit). 
This is handled as an error in CA compliance area and key agreement is not straightforward for RSA, therefore error seems like a valid option.

**Review by christopher-henderson — APPROVED** [review]

**Comment by christopher-henderson** [comment]
Thank you @mtgag! If you update the base branch then I can safely merge this (I could have sworn that GitHub gave me the button to do this as an admin? But maybe that's only for branches that originate from this repo...)

**Comment by christopher-henderson** [comment]
Hrmmm and yet GH still doesn't believe that the base branch is up-to-date. Considering that the only deletion from the PR is changing a copyright year I am going to go ahead and merge this.

**Comment by mtgag** [comment]
You may also close https://github.com/zmap/zlint/pull/599, because 599 has been incorporated in this PR.

**Pull request merged**